test/websocket/close.js

```diff
@@ -128,4 +130,53 @@ describe('Close', () => {
       ws.addEventListener('open', () => ws.close(3000))
     })
   })
+
+  test('calling close twice will only trigger the close event once', async (t) => {
+    t = tspl(t, { plan: 1 })
+
+    const server = new WebSocketServer({ port: 0 })
+
+    after(() => server.close())
+
+    server.on('connection', (ws) => {
+      ws.on('close', (code) => {
+        t.strictEqual(code, 1000)
+      })
+    })
+
+    const ws = new WebSocket(`ws://localhost:${server.address().port}`)
+    ws.addEventListener('open', () => {
+      ws.close(1000)
+      ws.close(1000)
+    })
+
+    await t.completed
+  })
+
+  test('calling close when connection is open but close handshake has started does not result in sending the handshake again', async (t) => {
```

Comment:
this test doesn't serve a purpose

Reply:
why not? This test covers the last else case

https://app.codecov.io/gh/nodejs/undici/pull/2865/blob/lib/web/websocket/websocket.js#L253

Reply:
because the test isn't 'natural', you can't mutate internal state normally. If a test could cover that with what websocket publicly exposes, it would be useful

Reply:
There is no way to simulate this with the used websocket server of the npm ws server.